### PR TITLE
Calendar Rendering and View Options

### PR DESCRIPTION
**Pull request by sunnybrie**
I've improved the calendar rendering. It's more precise now and shows 15-minute increments. 3-Day and Week views are also now available.

Upcoming:
 - Styling timeblocks to display information better on multi-day views.

### CODE REVIEW

File: app/src/App.css [hunk 1]
```diff
@@ -97,7 +97,7 @@ body::-webkit-scrollbar-thumb {
   height: 100%;
   width: 200px;
 
-  background-color: floralwhite;
+  background-color: ghostwhite;
```
Comment (carbonsoda) [inline app/src/App.css:100]: Would it be helpful to use variables instead? So that way it's just once place you have to change for all of them. 
```css
:root {
  --bg-color: ghostwhite;
}
...
{
  ...
  background-color: var(--bg-color);
  ...
}
```

File: app/src/components/NavBar.js [hunk 2]
```diff
@@ -19,4 +20,95 @@ function NavBar({ isAuthenticated, gcal, setTab, tab }) {
...
+  isAuthenticated,
+  gcal,
+  calView,
+  setCalView,
+  dateNavigation,
+  setDateNavigation,
+}) => {
+  const [gridColumn, setGridColumn] = useState("2 / auto");
+
+  //when user clicks "prev" or "next", tells dateNavigation how much to adjust the dates state by based on calView
```
Comment (carbonsoda) [inline app/src/components/NavBar.js:34]: Nice comment, it's direct and clear, makes it a lot easier to understand the rest of this 👌

File: app/src/components/TimeBlock.js [hunk 3]
```diff
@@ -1,27 +1,48 @@
...
+        return 2 + extra;
+      } else if (30 <= time.minute() < 45) {
+        return 3 + extra;
+      } else if (45 <= time.minute()) {
+        return 4 + extra;
+      }
+      return extra;
+    }
+
+    var blockStart =
```
Comment (carbonsoda) [inline app/src/components/TimeBlock.js:27]: Tiny thing, ES6 conventions prefers `let` or `const` instead of `var`.

File: app/src/components/TimeBlock.js [hunk 4]
```diff
@@ -1,27 +1,48 @@
 import React, { useState, useEffect } from "react";
 
+import dayjs from "dayjs";
+
 const TimeBlock = ({ start, end, summary, dayStart }) => {
+  const [gridRow, setGridRow] = useState("auto");
+  const [gridColumn, setGridColumn] = useState("auto / span 1");
+
   function getGridPlacement() {
```
Comment (carbonsoda) [inline app/src/components/TimeBlock.js:9]: This is a really neat method to figure out styling 👀

File: app/src/components/TimeLine.js [hunk 5]
```diff
@@ -1,18 +1,29 @@
...
   //Returns an array of time objects at half-hour increments with row location
   //based on dayStart starting hour and totalHours
   function getTimes() {
     let timesToRender = [];
-    for (let i = 0; i <= totalHours; i += 0.5) {
-      let timeString = `${new Date(dayStart).getHours() + Math.floor(i)}`;
-      if (i % 1 !== 0) {
-        timeString += ":30";
+    for (let i = 0; i <= totalHours; i += 0.25) {
+      let value = dayjs(dayStart).add(Math.floor(i), "hour"); //creates dayjs object, determines hour
```
Comment (carbonsoda) [inline app/src/components/TimeLine.js:11]: Maybe a more descriptive name would be better here, like `time`

File: app/src/components/TimeLine.js [hunk 6]
```diff
@@ -1,18 +1,29 @@
...
+      } else {
+        minute = 0;
       }
-      let timeLineLocation = `${i * 2 + 1}`;
+      value = value.set("minute", minute); //sets minute
+
+      let timeLineLocation = `${i * 4 + 1}`; //determines row to render on
 
-      timesToRender.push({ timeString, timeLineLocation });
+      timesToRender.push({ value, timeLineLocation }); //pushes as object
```
Comment (carbonsoda) [inline app/src/components/TimeLine.js:26]: The comment feels a lil extra/unneeded since the `{...}` already indicates a JS object.

File: server/db.mjs [hunk 7]
```diff
@@ -56,20 +56,40 @@ export const deleteTask = async (id) => {
...
 
 //USER QUERIES
 export const getUser = async (email) => {
-  const user = await db.any("SELECT * FROM users WHERE email = $1", [email]);
-  if (user.length < 1) {
+  const account = await db.any(
+    `SELECT * 
+FROM users
+JOIN settings
+  ON users."id" = settings."userId"
```
Comment (carbonsoda) [inline server/db.mjs:69]: Tiny thing, it might be better to shift `users."id"` to `users.id`, or consistently stick with one of those, of course unless it's case-sensitive like `"userId"`.

File: server/db.mjs [hunk 8]
```diff
@@ -56,20 +56,40 @@ export const deleteTask = async (id) => {
     [id],
   );
 
-  console.log({ deletedTask });
+  window.alert("Deleted Task: ", { deletedTask });
```
Comment (carbonsoda) [inline server/db.mjs:59]: Why did you change this from `console.log()` to `window.alert()`?

**Review by carbonsoda — COMMENTED** [review]
Generally things look pretty good! Looks like you did some refactoring too and clarifying comments. Just some very small extra refactoring, but they're pretty small.

I had an issue with adding tasks on the site, I'm unsure if that's intended at the moment, but just as a heads up if otherwise.

#### When I try to add a task I get a `503` error.

At first I accidentally did a time in the past, but even after correcting it to 7am tomorrow, the issue persisted.
- Initial time:
  - image
  - image

- After correcting the time:
  - image


When I removed the `min` from the time needed, so it said `estTime: 25` 
- image

When I changed it so `estTime: 2` 
- Looks like it posted! But issues with `GET`, so nothing shows up or updated on the calendar tab.
- image

File: app/src/App.css [hunk 9]
```diff
@@ -144,6 +143,67 @@ body::-webkit-scrollbar-thumb {
...
+.calendar .navTab.toggled {
+  z-index: 2;
+}
+
+.calendar .navTab.notToggled {
+  background-color: gainsboro;
+}
+
+/* TIMELINE */
+
```
Comment (yjwangit) [inline app/src/App.css:206]: Great job on the CSS styling!

File: app/src/components/Calendar.js [hunk 10]
```diff
@@ -1,61 +1,148 @@
...
+        break;
+      case "week":
+        getDates(7);
+      default:
+        break;
+    }
+  }
+
+  //USE EFFECTS
+
```
Comment (yjwangit) [inline app/src/components/Calendar.js:105]: Nice job on the ternary and switch usages!

File: app/src/components/NavBar.js [hunk 11]
```diff
@@ -19,4 +20,95 @@ function NavBar({ isAuthenticated, gcal, setTab, tab }) {
...
+        <span className="navTitle">Today</span>
+        <button className="next" onClick={() => prevNext(1)}>
+          next
+        </button>
+      </div>
+    </>
+  );
+};
+
 export default NavBar;
```
Comment (yjwangit) [inline app/src/components/NavBar.js:114]: Great job on the React useEffect and UI. Need to learn from yours.

**Review by yjwangit — APPROVED** [review]

File: app/src/App.css [hunk 12]
```diff
@@ -144,6 +143,67 @@ body::-webkit-scrollbar-thumb {
...
+  box-sizing: border-box;
+  position: relative;
+  top: 0em;
+  left: 0em;
+  margin: auto;
+  padding: 0em 0.125em;
+  width: 100%;
+  height: auto;
+
+  grid-template-columns: 1fr 3fr 1fr;
```
Comment (aedward8) [inline app/src/App.css:157]: Does this grid template layout mean you are using a grid CSS template from a 3rd party library?

File: app/src/App.css (hunk 9 above)
Comment (aedward8) [reply app/src/App.css:206]: I second this I will definitely be taking a peek later on to help me with my CSS file

File: app/src/App.js [hunk 13]
```diff
@@ -51,22 +51,34 @@ const App = () => {
...
         </div>
-      )}
+        <Planner {...{ isAuthenticated, gcal, timedownAccount }} />
+      </>
     </main>
   );
-};
+}
+
+function FooBar({ isAuthenticated, gcal }) {
```
Comment (aedward8) [inline app/src/App.js:74]: What does the fooBar function do?

File: app/src/components/Calendar.js [hunk 14]
```diff
@@ -1,61 +1,148 @@
...
+      let start = dayjs().hour(timedownAccount.sleepTime.start.hours);
+      let end = dayjs().hour(timedownAccount.sleepTime.end.hours);
+
+      setDayStart(start);
+      setDayEnd(end);
+
+      let total = dayjs(end).hour() - dayjs(start).hour();
+      // console.debug({ total }); //TEST
+      setTotalHours(total);
+    }
```
Comment (aedward8) [inline app/src/components/Calendar.js:42]: so cool that you added this customizable feature that pull from the user's settings! very thoughtful and considerate that people wake up and sleep at different times

File: app/src/components/Calendar.js [hunk 15]
```diff
@@ -1,61 +1,148 @@
...
+        ? "1fr 4fr"
+        : "1fr".concat(
+            newDates
+              .map((date, index) => {
+                return " 2fr";
+              })
+              .join(""),
+          );
+
+    console.debug("setting columns...", newDayColumns);
```
Comment (aedward8) [inline app/src/components/Calendar.js:83]: thats cool you are using console.debug!

File: server/db.mjs (hunk 8 above)
Comment (sunnybrie) [reply server/db.mjs:59]: Playing with the idea of a notifications system. So my thought process:  windows.alert() feels like a “sudo” notification so I thought I’d implement one to simulate and see how it feels..?

File: app/src/components/NavBar.js (hunk 11 above)
Comment (aedward8) [reply app/src/components/NavBar.js:114]: Yes you used UseEffect a lot and it works as intended

File: app/src/App.css (hunk 12 above)
Comment (sunnybrie) [reply app/src/App.css:157]: Nope! Just css from scratch. `fr` is just another unit of measurement that stands for _fluid ratio_. So it’s not flexible, but not based on percentages or surrounding elements or viewport, it’s just about maintaining the ratio between the elements in those columns. Super nifty.

File: app/src/App.js (hunk 13 above)
Comment (sunnybrie) [reply app/src/App.js:74]: Temp thing I was testing out with @richiethomas, at the time we couldn’t think of a name. He was suggesting I separate the logging-in and logging-out components. We used “Foobar” so I wouldn’t have to go through the time to change the already existing Login component.

[context1](https://github.com/sunnybrie/Timedown/pull/9#discussion_r630683877)
[context2](https://github.com/sunnybrie/Timedown/pull/9#discussion_r630685986)

**Review by aedward8 — APPROVED** [review]
Good Job Brie! I can see you have made so much progress in your project and the code logic looks pretty complex in certain parts!

File: app/src/App.css (hunk 1 above)
Comment (sunnybrie) [reply app/src/App.css:100]: I am so doing this.

File: app/src/App.js (hunk 13 above)
Comment (richiethomas) [reply app/src/App.js:74]: Actually I think there's a way to make this simpler than we did on the first pass.  Right now we've duplicated the wrapper divs like `main className="App"`.  As an alternative, a single return statement instead of the guard clauses could work like this:

```
  return (
    <main className="App">
      <>
        <div id="login">
          <Login {...{ isAuthenticated, gcal }} />
        </div>
        { isAuthenticated && <Planner {...{ isAuthenticated, gcal, timedownAccount }} />}
      </>
    </main>
  );
```

This would also remove the need for `UserDashboard` and `Foobar`. :-)

I do still think the `Logout` functionality should be extracted out of the `Login` component and into its own component for the sake of clarity.  But if `Login` is going to be returned either way then we *could* do the above instead. :-)

File: app/src/App.css (hunk 9 above)
Comment (carbonsoda) [reply app/src/App.css:206]: I third this, your styling work always looks so nice! 👀

File: server/db.mjs (hunk 8 above)
Comment (carbonsoda) [reply server/db.mjs:59]: Gotcha! I just was curious if it still shows up if the command is on the server side 😅 very neat if it does!!
